skfp/fingerprints/klekota_roth.py

```diff
@@ -4990,3 +181,82 @@ def transform(
             Array with fingerprints.
         """
         return super().transform(X, copy)
+
+    def _dict_to_node(self, d: dict) -> _PatternNode:
+        """
+        Recursively convert a dict representation of a pattern tree
+        into a _PatternNode tree.
+        """
+        node = _PatternNode()
+        node.smarts = d.get("smarts")
+        node.pattern_mol = Chem.MolFromSmarts(node.smarts) if node.smarts else None
+        node.is_terminal = d.get("is_terminal", False)
+        node.feature_bit = d.get("feature_bit")
+        node.atom_requirements = defaultdict(int, d.get("atom_requirements", {}))
+        node.children = [
+            self._dict_to_node(node_dict) for node_dict in d.get("children", [])
+        ]
+
+        if (
+            node.is_terminal
+            and node.smarts is not None
+            and node.feature_bit is not None
+        ):
+            self._feature_names[int(node.feature_bit)] = node.smarts
+        return node
+
+    def _load_tree(self) -> None:
+        """
+        Load the pattern tree from a JSON file into internal representation.
+        """
+        file = _TREE_PATH
+        if not file.exists():
+            raise FileNotFoundError(f"Tree file not found: {file}")
+
+        with file.open("r", encoding="utf-8") as f:
+            data = json.load(f)
+
+        for key in ("n_nodes", "atoms", "tree"):
+            if key not in data:
+                raise KeyError(f"Missing key {key} in tree file {file}")
+
+        self._feature_names = [""] * data["n_terminal_nodes"]
+        self._pattern_atoms = {key: Chem.MolFromSmarts(key) for key in data["atoms"]}
+        self._root = self._dict_to_node(data["tree"])
+
+        if any(f == "" for f in self._feature_names):
+            raise ValueError("SMARTS or feature_bit missing in terminal nodes")
+
+    def _calculate_fingerprint(self, X: Sequence[str | Mol]) -> np.ndarray | csr_array:
+        X = ensure_mols(X)
+
+        n_bits = self.n_features_out
+        bits = np.zeros((len(X), n_bits), dtype=np.uint32 if self.count else np.uint8)
+        root_children = self._root.children
+
+        if self.count:
+            set_value = lambda mol, pattern: len(mol.GetSubstructMatches(pattern))
+        else:
+            set_value = lambda _mol, _pattern: 1
+
+        for i, mol in enumerate(X):
+            stack: deque[_PatternNode] = deque(root_children)
+            atom_contents = defaultdict(int)
+            for key, atom in self._pattern_atoms.items():
+                atom_contents[key] = len(mol.GetSubstructMatches(atom))
```

Comment:
For checking atom types, iteration over atoms should be faster than individual SMARTS patterns